// @smartface/contx/lib/smartface/componentContextPatch.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(["module", "exports", "./pageContext", "../util/patchMethod", "sf-core/application"], factory);
  } else if (typeof exports !== "undefined") {
    factory(module, exports, require("./pageContext"), require("../util/patchMethod"), require("sf-core/application"));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod, mod.exports, global.pageContext, global.patchMethod, global.application);
    global.componentContextPatch = mod.exports;
  }
})(this, function (module, exports, _pageContext, _patchMethod, Application) {
  "use strict";

  Object.defineProperty(exports, "__esModule", {
    value: true
  });
  exports.default = componentContextPatch;

  var _pageContext2 = _interopRequireDefault(_pageContext);

  var _patchMethod2 = _interopRequireDefault(_patchMethod);

  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
      default: obj
    };
  }

  function componentContextPatch(component, name) {
    component.themeContext = Application.theme();
    component.themeContext({
      type: "addThemeableContext",
      name: name,
      pageContext: (0, _pageContext2.default)(component, name, null, null)
    });
    component.onShow = (0, _patchMethod2.default)(component, "onShow", onShow);

    function onShow(superOnShow, params) {
      superOnShow(params);
    }
  }
  module.exports = exports["default"];
});
