// @smartface/contx/lib/smartface/index.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(["module", "exports", "./action/addChild", "./action/removeChild", "./action/removeChildren", "./pageContext", "./pageContextPatch", "./componentContextPatch"], factory);
  } else if (typeof exports !== "undefined") {
    factory(module, exports, require("./action/addChild"), require("./action/removeChild"), require("./action/removeChildren"), require("./pageContext"), require("./pageContextPatch"), require("./componentContextPatch"));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod, mod.exports, global.addChild, global.removeChild, global.removeChildren, global.pageContext, global.pageContextPatch, global.componentContextPatch);
    global.index = mod.exports;
  }
})(this, function (module, exports, _addChild, _removeChild, _removeChildren, _pageContext, _pageContextPatch, _componentContextPatch) {
  "use strict";

  Object.defineProperty(exports, "__esModule", {
    value: true
  });

  var _addChild2 = _interopRequireDefault(_addChild);

  var _removeChild2 = _interopRequireDefault(_removeChild);

  var _removeChildren2 = _interopRequireDefault(_removeChildren);

  var _pageContext2 = _interopRequireDefault(_pageContext);

  var _pageContextPatch2 = _interopRequireDefault(_pageContextPatch);

  var _componentContextPatch2 = _interopRequireDefault(_componentContextPatch);

  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
      default: obj
    };
  }

  exports.default = {
    addChild: _addChild2.default,
    removeChild: _removeChild2.default,
    removeChildren: _removeChildren2.default,
    pageContext: _pageContext2.default,
    pageContextPatch: _pageContextPatch2.default,
    componentContextPatch: _componentContextPatch2.default
  };
  module.exports = exports["default"];
});
