// @smartface/contx/lib/styling/index.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(["module", "exports", "action/addThemeableContext", "action/changeTheme", "action/pushClassNames", "action/removeClassNames", "action/updateContextTree", "./ThemeContext", "./StyleContext"], factory);
  } else if (typeof exports !== "undefined") {
    factory(module, exports, require("action/addThemeableContext"), require("action/changeTheme"), require("action/pushClassNames"), require("action/removeClassNames"), require("action/updateContextTree"), require("./ThemeContext"), require("./StyleContext"));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod, mod.exports, global.addThemeableContext, global.changeTheme, global.pushClassNames, global.removeClassNames, global.updateContextTree, global.ThemeContext, global.StyleContext);
    global.index = mod.exports;
  }
})(this, function (module, exports, _addThemeableContext, _changeTheme, _pushClassNames, _removeClassNames, _updateContextTree, _ThemeContext, _StyleContext) {
  "use strict";

  Object.defineProperty(exports, "__esModule", {
    value: true
  });

  var _addThemeableContext2 = _interopRequireDefault(_addThemeableContext);

  var _changeTheme2 = _interopRequireDefault(_changeTheme);

  var _pushClassNames2 = _interopRequireDefault(_pushClassNames);

  var _removeClassNames2 = _interopRequireDefault(_removeClassNames);

  var _updateContextTree2 = _interopRequireDefault(_updateContextTree);

  var _ThemeContext2 = _interopRequireDefault(_ThemeContext);

  var _StyleContext2 = _interopRequireDefault(_StyleContext);

  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
      default: obj
    };
  }

  exports.default = {
    addThemeableContext: _addThemeableContext2.default,
    changeTheme: _changeTheme2.default,
    pushClassNames: _pushClassNames2.default,
    removeClassName: _removeClassNames2.default,
    removeClassNames: _removeClassNames2.default,
    updateContextTree: _updateContextTree2.default,
    ThemeContext: _ThemeContext2.default,
    StyleContext: _StyleContext2.default
  };
  module.exports = exports["default"];
});
